plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -0,0 +1,398 @@
+to that subterm (not including itself).
+
+In the above example, the ID of `\x` is 0, the ID of `\y` is 1, and the IDs of the
+three case branches are 2, 3, 4 (the actual numbers don't matter, as long as they are unique).
+The path for the first `1+(2+x)` and the first `2+x` is "0.1"; the path for the second
+`1+(2+x)` and the second `2+x` is "0.1.2"; the path for `4+x` is "0.1.4".
+
+In the third pass, we calculate a count for each `(term, path)` pair, where `term` is a
+non-workfree term, and `path` is its path. If the same term has two paths, and one is an
+ancestor (i.e., prefix) of the other, we increment the count for the ancestor path in both
+instances.
+
+In the above example, there are three occurrences of `2+x`, whose paths are "0.1", "0.1.2"
+and "0.1.3", respectively. The first path is an ancestor of the latter two. Therefore,
+the count for `(2+x, "0.1")` is 3, while the count for `(2+x, "0.1.2")` and `(2+x, "0.1.3")`
+is 0. The following all have a count of 1: `(3+x, "0.1.2")`, `(3+x, "0.1.3")` and
+`(4+x, "0.1.4")`.
+
+Now, each `(term, path)` pair whose count is greater than 1 is a CSE candidate.
+In the above example, the CSE candidates are `(2+x, "0.1")` and `(1+(2+x), "0.1")`.
+Note that `3+x` is not a CSE candidate, because it has two paths, and neither has a count
+greater than 1. `2+` is also not a CSE candidate, because it is workfree.
+
+The CSE candidates are then processed in descending order of their `termSize`s. For each CSE
+candidate, we generate a fresh variable, create a LamAbs for it under its path, and substitute
+it for all occurrences in the original term whose paths are descendents (or self) of
+the candidate's path. The order is because a bigger expression may contain a small subexpression.
+
+In the above example, we first process CSE candidate `(1+(2+x), "0.1")`. We create a fresh
+variable `cse1` for it, perform substitution, and create a `LamAbs` under path "0.1" (i.e., around
+the body of `y`). After processing this CSE candidate, the original term becomes
+
+\x y -> (\cse1 -> cse1
+                  +
+                  (case y [ cse1 + (3+x)
+                          , (2+x) + (3+x)
+                          , 4+x
+                          ]
+        ) (1+(2+x))
+
+The second CSE candidate is processed similarly, and the final result is
+
+\x y -> (\cse2 -> (\cse1 -> cse1
+                            +
+                            (case y [ cse1 + (3+x)
+                                    , cse2 + (3+x)
+                                    , 4+x
+                                    ]
+                  ) (1+cse2)
+        ) (2+x)
+
+Here's another example:
+
+force (force ifThenElse
+         (lessThanEqualsInteger 0 0)
+         (delay ((1+2) + (1+2)))
+         (delay (1+2))
+      )
+
+In this case, the first two occurrences of `1+2` can be CSE'd, but the third occurrence
+can not. This is ensured by checking the path when substituting `cse1` for `1+2`. The result is
```

Comment:
Why not? I think the answer is:
1. There is no scope enclosing both of them? If so, this is probably a bug, there should be a scope for "the top level" for sure.
2. They have differing paths, corresponding to the two different delays. In this instance, since the scopes come from `delay`s, you _could_ share the value across both (since there isn't a newly introduced variable that they might depend on). However, that might lead to the value being computed sooner than it would otherwise have been, which seems bad.

So, not at all obvious. And I don't know how to reason about it because you haven't told me what your algorithm is trying to do!

Reply:
> However, that might lead to the value being computed sooner than it would otherwise have been, which seems bad.

Ok, this is not a great example because (1+2) appears in both branches, so it will definitely be computed. A better example would be an expression that appears in two of the three case branches, in which case we definitely can't CSE it.

Reply:
I wasn't assuming we knew it would definitely be computed, just that we could lift it anyway (and then get computation too soon).